src/ffi/cpython/unicodeobject.rs

```diff
@@ -0,0 +1,314 @@
+use crate::ffi::{PyObject, PyUnicode_Check, Py_UCS1, Py_UCS2, Py_UCS4, Py_hash_t, Py_ssize_t};
+use libc::wchar_t;
+use std::os::raw::{c_char, c_int, c_uint, c_void};
+
+// skipped Py_UNICODE_ISSPACE()
+// skipped Py_UNICODE_ISLOWER()
+// skipped Py_UNICODE_ISUPPER()
+// skipped Py_UNICODE_ISTITLE()
+// skipped Py_UNICODE_ISLINEBREAK
+// skipped Py_UNICODE_TOLOWER
+// skipped Py_UNICODE_TOUPPER
+// skipped Py_UNICODE_TOTITLE
+// skipped Py_UNICODE_ISDECIMAL
+// skipped Py_UNICODE_ISDIGIT
+// skipped Py_UNICODE_ISNUMERIC
+// skipped Py_UNICODE_ISPRINTABLE
+// skipped Py_UNICODE_TODECIMAL
+// skipped Py_UNICODE_TODIGIT
+// skipped Py_UNICODE_TONUMERIC
+// skipped Py_UNICODE_ISALPHA
+// skipped Py_UNICODE_ISALNUM
+// skipped Py_UNICODE_COPY
+// skipped Py_UNICODE_FILL
+// skipped Py_UNICODE_IS_SURROGATE
+// skipped Py_UNICODE_IS_HIGH_SURROGATE
+// skipped Py_UNICODE_IS_LOW_SURROGATE
+// skipped Py_UNICODE_JOIN_SURROGATES
+// skipped Py_UNICODE_HIGH_SURROGATE
+// skipped Py_UNICODE_LOW_SURROGATE
+
+// This type is defined inline in CPython source.
+#[repr(C)]
+pub struct PyASCIIObjectState {
+    // SSTATE_* constants.
+    pub interned: c_uint,
+    // PyUnicode_*_KIND constants.
+    pub kind: c_uint,
+    pub compact: c_uint,
+    pub ascii: c_uint,
+    pub ready: c_uint,
+    pub _a: c_uint,
+}
```

Comment:
This struct looks wrong; the upstream CPython source implements this state as a 32-bit [bit field](https://www.tutorialspoint.com/cprogramming/c_bit_fields.htm).

It's not possible to have an equivalent bitfield definition in Rust (e.g. see https://github.com/rust-lang/rfcs/issues/314)

Looks like the `cpython` crate defines the `state` member as a single `u32` field and uses bitwise operators to read values at various offsets. That seems to make sense, but as far as I'm aware the bitfield order isn't actually well-defined (https://stackoverflow.com/a/19376742) so that seems like it's actually a bug. I guess it'll work on some platforms and be UB on others?

I think some more research is needed on the correct way to handle this.